client/cli/derive/src/substrate_cli.rs

```diff
@@ -0,0 +1,175 @@
+	let attrs = match Punctuated::<ExprAssign, Token![,]>::parse_terminated.parse(a) {
+		Ok(x) => x,
+		Err(err) => abort!(err.span(), "could not parse attributes: {}", err),
+	};
+
+	let mut attrs = attrs
+		.iter()
+		.map(|x| {
+			(
+				match &*x.left {
+					Expr::Path(p) if p.path.segments.len() == 1 => {
+						p.path.segments[0].ident.to_string()
+					}
+					_ => abort!(x.left.span(), "could not parse token"),
+				},
+				(*x.right).to_token_stream(),
+			)
+		})
+		.collect::<HashMap<_, _>>();
+
+	let pkg_name = std::env::var("CARGO_PKG_NAME").unwrap().to_token_stream();
+	let impl_name = attrs
+		.remove("impl_name")
+		.unwrap_or_else(|| pkg_name.clone());
+	let impl_version = attrs
+		.remove("impl_version")
+		.unwrap_or_else(|| get_version().to_token_stream());
+	let support_url = attrs
+		.remove("support_url")
+		.unwrap_or_else(|| abort_call_site!("missing attribute: support_url"));
+	let executable_name = attrs
+		.remove("executable_name")
+		.unwrap_or_else(|| pkg_name.clone());
+	let author = attrs.remove("author").unwrap_or_else(|| {
+		std::env::var("CARGO_PKG_AUTHORS")
+			.unwrap_or_default()
+			.to_token_stream()
+	});
+	let description = attrs.remove("description").unwrap_or_else(|| {
+		std::env::var("CARGO_PKG_DESCRIPTION")
+			.unwrap_or_default()
+			.to_token_stream()
+	});
+	let copyright_start_year = attrs
+		.remove("copyright_start_year")
+		.unwrap_or_else(|| abort_call_site!("missing attribute: copyright_start_year"));
+
+	if !attrs.is_empty() {
+		abort_call_site!(
+			"unknown macro parameters: {}",
+			attrs
+				.keys()
+				.map(|x| x.as_str())
+				.collect::<Vec<_>>()
+				.join(", ")
+		);
+	}
+
+	let mut i: ItemImpl = match syn::parse(i) {
+		Ok(x) => x,
+		_ => abort_call_site!("this macro only works on an impl"),
+	};
+
+	let existing_methods = i
+		.items
+		.iter()
+		.filter_map(|x| match x {
+			ImplItem::Method(x) => Some(x.sig.ident.to_string()),
+			_ => None,
+		})
+		.collect::<HashSet<_>>();
+
+	define_if_missing!(existing_methods, i, {
+		fn impl_name() -> &'static str
+	});
+
+	define_if_missing!(existing_methods, i, {
+		fn impl_version() -> &'static str
+	});
+
+	define_if_missing!(existing_methods, i, {
+		fn support_url() -> &'static str
+	});
+
+	define_if_missing!(existing_methods, i, {
+		fn executable_name() -> &'static str
+	});
+
+	define_if_missing!(existing_methods, i, {
+		fn author() -> &'static str
+	});
+
+	define_if_missing!(existing_methods, i, {
+		fn description() -> &'static str
+	});
+
+	define_if_missing!(existing_methods, i, {
+		fn copyright_start_year() -> i32
+	});
+
+	quote!(#i)
+}
+
+fn get_platform() -> String {
+	use platforms::*;
+
+	let env_dash = if TARGET_ENV.is_some() { "-" } else { "" };
+
+	format!(
+		"{}-{}{}{}",
+		TARGET_ARCH.as_str(),
+		TARGET_OS.as_str(),
+		env_dash,
+		TARGET_ENV.map(|x| x.as_str()).unwrap_or(""),
+	)
+}
+
+fn get_version() -> String {
+	let impl_commit = std::env::var("VERGEN_SHA_SHORT")
```

Comment:
It is not guaranteed that the macro is called when this env variable changes!

Reply:
The macro is used on the binary/cli. The binary depends of all its parts (service, etc...). I don't see how it wouldn't be re-triggered, it's on the top of the build. Can you find an example?

Reply:
Just because the build is triggered, doesn't mean the macro is called again. Rustc caches compilation outputs to not rebuild stuff every time.

Reply:
Does `rerun_if_git_head_changed` force the rebuild? https://docs.rs/substrate-build-script-utils/2.0.0-alpha.3/substrate_build_script_utils/fn.rerun_if_git_head_changed.html

Reply:
That just makes sure that the `build` script is executed again.

Reply:
Well I don't know. Can you show me an example or some doc? I'm not sure I understand what is the problem

Reply:
(I do it exactly like they do in structopt: https://github.com/TeXitoi/structopt/blob/master/structopt-derive/src/attrs.rs#L330 )

Reply:
https://users.rust-lang.org/t/equivalent-of-cargo-rerun-if-changed-for-procedural-macros/37036 for example. Nothing says that the macro is re-run.

Reply:
The macro won't rerun because vergen_sha_short has changed indeed, but this is fine.

The code in the macro is executed during the build of the binary. The binary is on top of everything so the code will be always executed. This works exactly the same than structopt works. That particular macro (substrate_cli) is only used in the binary, not in sc-cli.